### PR TITLE
Sync ReSDK with latest Resolwe

### PR DESCRIPTION
**Pull request by JureZmrzlikar**

### CODE REVIEW

**Comment by JureZmrzlikar** [comment]
@dblenkus , @mstajdohar please review. All tests pass locally, Jenkins tests are failing due to not-up-to-date `genialis-base`.

File: resdk/resources/data.py [hunk 1]
```diff
@@ -28,27 +28,34 @@ class Data(BaseResolweResource):
...
-    _collections = None
-
-    WRITABLE_FIELDS = ('descriptor_schema', 'descriptor',
-                       'tags') + BaseResolweResource.WRITABLE_FIELDS
-    UPDATE_PROTECTED_FIELDS = ('input', 'process') + BaseResolweResource.UPDATE_PROTECTED_FIELDS
-    READ_ONLY_FIELDS = ('process_input_schema', 'process_output_schema', 'output', 'started',
-                        'finished', 'checksum', 'status', 'process_progress', 'process_rc',
-                        'process_info', 'process_warning', 'process_error', 'process_type',
-                        'process_name') + BaseResolweResource.READ_ONLY_FIELDS
+    READ_ONLY_FIELDS = BaseResolweResource.READ_ONLY_FIELDS + (
```
Comment (mstajdohar) [inline resdk/resources/data.py]: Add a test that checks that fields in READ_ONLY_FIELDS, UPDATE_PROTECTED_FIELDS and WRITABLE_FIELDS are defined in the resource classes.

File: resdk/shortcuts/collection.py [hunk 2]
```diff
@@ -1,381 +1,100 @@
...
             raise ValueError("`positions` argument must be list.")
 
-        if positions:
-            if len(samples) != len(positions):
-                raise ValueError("`samples` and `positions` arguments must be of the same length.")
+        if not isinstance(labels, list):
+            raise ValueError("`labels` argument must be list.")
+
+        if positions and labels and not len(samples) == len(positions) == len(labels):
+            msg = "`samples`, `positions` and `labels` arguments must be of the same length."
```
Comment (dblenkus) [inline resdk/shortcuts/collection.py]: Why don't you insert message directly into the exception?

File: resdk/resources/sample.py [hunk 3]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+        elif self_cases > 1:
+            msg = ('This sample is a case in a background relation with also other samples as '
+                   'cases. If you would like to change background relation for all of them please '
+                   'delete current relation and create new one with desired background.')
+            raise ValueError(msg)
+
+        self.save()
+        self.update()
+
+    def is_background(self):
```
Comment (dblenkus) [inline resdk/resources/sample.py]: What is the difference between `Sample.is_background` and `Sample.background`?

File: resdk/resources/sample.py [hunk 4]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+        # to check that both, background and case are defined in only
+        # one common collection. Actions are done on this collection.
+        common_ids = {col.id for col in self.collections} & {col.id for col in bground.collections}
+        if not common_ids:
+            raise ValueError('Background and case sample are not in the same collection.')
+        elif len(common_ids) > 1:
+            raise ValueError('Background and case sample are in multiple common collections.')
+        collection = self.resolwe.collection.get(list(common_ids)[0])
+
+        def count_cases(entity, label):
```
Comment (dblenkus) [inline resdk/resources/sample.py]: Define function at the top of the outer function.

File: resdk/resources/sample.py [hunk 5]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+            Relation is specified by collection, type-background'entity and label.
+            """
+            relation = list(self.resolwe.relation.filter(
+                collection=collection.id,
+                type='background',
+                entity=entity.id,
+                label=label,
+            ))
+            if len(relation) > 1:
+                raise ValueError('Multiple relations of type "background" for sample {} in '
```
Comment (dblenkus) [inline resdk/resources/sample.py]: New line after opening bracket and before closing one.

File: resdk/resources/sample.py [hunk 6]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+        elif self_cases == 1 and bground_cases == 0:
+            for partition in self_relation.partitions:
+                if partition['label'] == 'background':
+                    partition['entity'] = bground.id
+                    break
+        elif self_cases == 1 and bground_cases > 0:
+            self_relation.delete(force=True)
+            bground_relation.add_sample(self, label='case')
+        elif self_cases > 1:
+            msg = ('This sample is a case in a background relation with also other samples as '
```
Comment (dblenkus) [inline resdk/resources/sample.py]: New line after opening bracket and before closing one.

File: resdk/resources/sample.py [hunk 7]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+            self._background = self._get_background(fail_silently=True)
+            if self._background is None:
+                # Cache the result = no background is found.
+                self._background = False
+
+        if self._background:
+            return self._background
+
+    @background.setter
+    def background(self, bground):
```
Comment (dblenkus) [inline resdk/resources/sample.py]: Is all this logic necessary here? Why don't you move it to the backend?

File: resdk/resources/data.py (hunk 1 above)
Comment (JureZmrzlikar) [reply resdk/resources/data.py]: Done.

File: resdk/resources/relation.py [hunk 8]
```diff
@@ -105,19 +96,22 @@ def update(self):
...
         super(Relation, self).update()
 
-    def add_sample(self, sample, position=None):
-        """Add ``sample`` object to the collection."""
-        payload = [{
-            'entity': get_sample_id(sample),
-            'position': position
-        }]
-        self.api(self.id).add_entity.post(payload)
+    def add_sample(self, sample, label=None, position=None):
```
[the remark this answers is not in the document]
Comment (JureZmrzlikar) [reply resdk/resources/relation.py]: No. Adding a sample to relation modifies `RealtionPartition` and it still has `label` and `position` attributes. `category` attribute is attribute of `Relation` object.

File: resdk/resources/sample.py [hunk 9]
```diff
@@ -126,27 +142,116 @@ def confirm_is_annotated(self):
...
+            self_relation.delete(force=True)
+            bground_relation.add_sample(self, label='case')
+        elif self_cases > 1:
+            msg = ('This sample is a case in a background relation with also other samples as '
+                   'cases. If you would like to change background relation for all of them please '
+                   'delete current relation and create new one with desired background.')
+            raise ValueError(msg)
+
+        self.save()
+        self.update()
```
[the remark this answers is not in the document]
Comment (JureZmrzlikar) [reply resdk/resources/sample.py]: Good point, will fix.

**Comment by JureZmrzlikar** [comment]
@mstajdohar I would like to adress tutorials in a separate PR. This one is really huge already.

**Comment by mstajdohar** [comment]
Go ahead and merge.

**Pull request merged**